### ISSUE TITLE
endpoint: stdio

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4697]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)| @glyph reported|
|-|-|
|Trac ID|trac#4697|
|Type|enhancement|
|Created|2010-10-19 01:28:14Z|
|Branch|https://github.com/twisted/twisted/tree/stdio-endpoint-4697-2|

You should be able to run a Twisted server via a facility like inetd, or just as a subprocess for debugging.

This is a little tricky, as there can only be one, and it will interact with daemonization and other implementation details, and may need to have some nasty clever bits in order to stop the reactor when the connection is done.

I think this should be invoked as something like: `serverFromString("stdio")`.

<details><summary>Searchable metadata</summary>

```
trac-id__4697 4697
type__enhancement enhancement
reporter__glyph glyph
priority__normal normal
milestone__ 
branch__branches_stdio_endpoint_4697_2 branches/stdio-endpoint-4697-2
branch_author__ashfall ashfall
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1287451694000000 1287451694000000
changetime__1340900539000000 1340900539000000
version__None None
owner__ashfall ashfall
cc__itamarst
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_2"></a>|@itamarst commented|
|-|-|

inetd is not the only use case; one might want to e.g. run a pipeline tool (assuming ticket [[#3442](https://github.com/twisted/twisted/issues/3442)](https://github.com/twisted/twisted/issues/3442) is fixed). #3442 might actually affect the inetd case too, actually...

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_4"></a>|@glyph commented|
|-|-|

Replying to [itamar](#note_2):
> inetd is not the only use case; one might want to e.g. run a pipeline tool

absolutely!

> (assuming ticket [[[#3442](https://github.com/twisted/twisted/issues/3442)](https://github.com/twisted/twisted/issues/3442)](https://github.com/twisted/twisted/issues/3442) is fixed). #3442 might actually affect the inetd case too, actually...

Actually, this is totally orthogonal.  #3442 is about _sharing_ the _same_ file description with other processes, as with a controlling terminal.  But the read end and the write end of a pipe are distinct file descriptions, so you can mess with your end without disturbing your peer's view of the socket or pipe.  In the inetd case, the file descriptor is all yours (plus, in that case, it's a socket, and the other end is quite possibly in somebody else's kernel).

But even in the local case the flags are independent:

```
>>> from os import O_NONBLOCK, pipe
>>> from fcntl import F_GETFL, F_SETFL, fcntl
>>> pipe()
(3, 4)
>>> fcntl(3, F_GETFL)
0
>>> fcntl(4, F_GETFL)
1
>>> fcntl(3, F_SETFL, O_NONBLOCK)
0
>>> fcntl(3, F_GETFL)
4
>>> fcntl(4, F_GETFL)
1
```

QED.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="ashfall's avatar" src="https://avatars.githubusercontent.com/u/1138873?s=50" width="50" height="50">](https://github.com/ashfall)|@ashfall set owner to @ashfall|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="ashfall's avatar" src="https://avatars.githubusercontent.com/u/1138873?s=50" width="50" height="50">](https://github.com/ashfall)<a name="note_7"></a>|@ashfall commented|
|-|-|

(In [34517]) Branching to 'stdio-endpoint-4697'

**Comment by twisted-trac** [comment]
|[<img alt="ashfall's avatar" src="https://avatars.githubusercontent.com/u/1138873?s=50" width="50" height="50">](https://github.com/ashfall)|@ashfall removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="ashfall's avatar" src="https://avatars.githubusercontent.com/u/1138873?s=50" width="50" height="50">](https://github.com/ashfall)<a name="note_9"></a>|@ashfall commented|
|-|-|

[Build Results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/stdio-endpoint-4697)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="ashfall's avatar" src="https://avatars.githubusercontent.com/u/1138873?s=50" width="50" height="50">](https://github.com/ashfall)<a name="note_11"></a>|@ashfall commented|
|-|-|

(In [34615]) Branching to 'stdio-endpoint-4697-2'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun set owner to @ashfall<br>@exarkun set status to `new`|
|-|-|

Thanks.  This looks pretty nice.  Just a few minor points:

  1. The `stdioProtocol` attribute of the endpoint doesn't look like it's needed.  It's also not part of the endpoint interface, so code that works with arbitrary endpoints won't be able to use it (in other words, no other endpoints have an attribute with that name).
  1. The test method docstrings should be a little more detailed about what's being tested, and not start with "Test to check".  Document test methods like "The reactor passed to the endpoint is set as its _reactor attribute."
  1. There should also be a test verifying that a `PipeAddress` instance is passed to `buildProtocol`.  I suggest doing this with a custom factory that records the address passed to it on an attribute of itself and then using `assertIsInstance` in the test to verify that object was an instance of `PipeAddress`.
  1. It looks like Windows has some difficulty with this code.  It may be necessary to move the definition of `PipeAddress` out of `_process.py`, since that module is not importable on Windows (check out the build results on the Windows slaves for details).

Thanks again!

**Comment by twisted-trac** [comment]
|[<img alt="ashfall's avatar" src="https://avatars.githubusercontent.com/u/1138873?s=50" width="50" height="50">](https://github.com/ashfall)<a name="note_13"></a>|@ashfall removed owner|
|-|-|

Fixed.
(Hopefully the build failures as well)

[Build Results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/ipv6-tcp-client-endpoint-5695-2)

**Comment by twisted-trac** [comment]
|[<img alt="ashfall's avatar" src="https://avatars.githubusercontent.com/u/1138873?s=50" width="50" height="50">](https://github.com/ashfall)<a name="note_14"></a>|@ashfall commented|
|-|-|

I pasted build results for a wrong branch in the above comment, sorry about the same.
Here's the appropriate link: 
[Build Results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/stdio-endpoint-4697-2)

**Comment by twisted-trac** [comment]
|[<img alt="ashfall's avatar" src="https://avatars.githubusercontent.com/u/1138873?s=50" width="50" height="50">](https://github.com/ashfall)<a name="note_15"></a>|@ashfall set owner to @ashfall|
|-|-|

Found more errors, working on it.

**Comment by twisted-trac** [comment]
|[<img alt="ashfall's avatar" src="https://avatars.githubusercontent.com/u/1138873?s=50" width="50" height="50">](https://github.com/ashfall)<a name="note_16"></a>|@ashfall removed owner|
|-|-|

Looks better now. Sorry for all the mess.

[Build Results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/stdio-endpoint-4697-2)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_17"></a>|@exarkun set owner to @ashfall|
|-|-|

Thanks.

  1. in `twisted/internet/test/test_endpoints.py`, `stdioFactory` is misnamed according to the naming convention.  Classes are named like `StdioFactory`.
  1. Also don't write "should" in a test method docstring.
  1. The docstring for `test_protocol` talks about something not very interesting.  The test isn't a test for factories.  It's a test for the stdio endpoint.  And in particular, it's a test for the stdio endpoint using the factory to get the protocol instance to use.  So the docstring should talk about that instead.  We already know that `Factory.buildProtocol` instantiates whatever the factory's `protocol` attribute is set to.
  1. There are some reactor dirty warnings when the tests run on Windows.  This is caused by only calling `loseConnection`, not waiting for the reactor to finish disconnecting the stdio connection.  There are two ways to approach solving this.  One would be to use a fake reactor that doesn't really establish a stdio connection.  The other would be to wait for `connectionLost` to be called on the protocol before firing the Deferred returned to trial.  If you'd rather defer this fix to another ticket, I think that's fine (just please file that ticket when you close this one).
  1. The `Win32PipeAddress` subclass in `twisted/internet/stdio.py` is unnecessary.  Also, I don't _quite_ understand how `test_address` passes with the implementation working this way. :)  You can provide a new name for an import in a couple ways.  Either `PipeAddress = Win32Address` or by changing the import to `from ... import Win32PipeAddress as PipeAddress`.  This avoids creating a new class, which is wrong, because the Windows stdio implementation won't actually use the new class, it'll keep using the base class.
  1. There is one pyflakes warning for endpoints.py and one pyflakes warning for stdio.py

Thanks again

**Comment by twisted-trac** [comment]
|[<img alt="ashfall's avatar" src="https://avatars.githubusercontent.com/u/1138873?s=50" width="50" height="50">](https://github.com/ashfall)<a name="note_18"></a>|@ashfall removed owner|
|-|-|

Replying to [exarkun](#note_17):

>   1. There are some reactor dirty warnings when the tests run on Windows.  This is caused by only calling `loseConnection`, not waiting for the reactor to finish disconnecting the stdio connection.  There are two ways to approach solving this.  One would be to use a fake reactor that doesn't really establish a stdio connection.  The other would be to wait for `connectionLost` to be called on the protocol before firing the Deferred returned to trial.  If you'd rather defer this fix to another ticket, I think that's fine (just please file that ticket when you close this one).

I'll file a new ticket once this is ready to merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_19"></a>|@exarkun set owner to @ashfall|
|-|-|

Thanks!  A couple more tiny points:

  1. There's a pyflakes warning in `test_endpoints.py`
  1. `StandardIOEndpoint.listen`'s docstring appears to be truncated :)  Please make it a bit more specific, like the docstrings for other `listen` implementations.

Once those are addressed, please merge.  Thanks again.

**Comment by twisted-trac** [comment]
|[<img alt="ashfall's avatar" src="https://avatars.githubusercontent.com/u/1138873?s=50" width="50" height="50">](https://github.com/ashfall)<a name="note_20"></a>|@ashfall commented|
|-|-|

Filed [#5629](https://github.com/twisted/twisted/issues/5629) for the string description plugin, and [#5630](https://github.com/twisted/twisted/issues/5630) for the dirty reactor warnings on Windows.

**Comment by twisted-trac** [comment]
|[<img alt="ashfall's avatar" src="https://avatars.githubusercontent.com/u/1138873?s=50" width="50" height="50">](https://github.com/ashfall)<a name="note_21"></a>|@ashfall set status to `closed`|
|-|-|

(In [34653]) Merge stdio-endpoint-4697-2: An endpoint interface that listens on stdin/stdout

Author: ashfall
Reviewer: exarkun
Fixes: [#4697](https://github.com/twisted/twisted/issues/4697)

Added StandardIOEndpoint to twisted/internet/endpoints.py, along with corresponding tests in twisted/internet/test/test_endpoints.py

**Comment by twisted-trac** [comment]
|[<img alt="ashfall's avatar" src="https://avatars.githubusercontent.com/u/1138873?s=50" width="50" height="50">](https://github.com/ashfall)<a name="note_22"></a>|@ashfall commented|
|-|-|

Replying to [ashfall](#note_20):
> Filed [#5629](https://github.com/twisted/twisted/issues/5629) for the string description plugin, and [#5630](https://github.com/twisted/twisted/issues/5630) for the dirty reactor warnings on Windows.

Err, I mean [#5729](https://github.com/twisted/twisted/issues/5729) and [#5730](https://github.com/twisted/twisted/issues/5730) respectively.
